ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -510,6 +502,14 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = do
 
     processLoE curChain (LgrDB.ledgerDbCurrent ledgerDB) (writeTVar cdbLoEFrag) cdbLoE
 
+    loeFrag <- atomically (cross curChain <$> readTVar cdbLoEFrag) <&> \case
+        Just (_, frag) -> frag
+        -- We don't crash if the LoE fragment doesn't intersect with the selection
+        -- because we update the selection _after_ updating the LoE fragment, which
+        -- means it could move to another fork or beyond the end of the LF, depending
+        -- on the implementation of @updateLoEFrag@.
+        Nothing        -> AF.Empty (AF.anchor curChain)
```

Comment:
It looks to me like `computeLoEMaxExtra` below should produce `loeLimit + however many blocks are needed to reach the tip of loeFrag`. However, it is computed instead as `loeLimit - budgetAlreadyUsed`.

Reply:
Are you just suggesting a refactoring, or a logic change? In case of the latter: Can you give a concrete example of a situation you think this rule would handle better/correctly in contrast to the existing one?

Reply:
Suppose `loeFrag`'s tip is the block following `hdr` (the block being added). The current implementation will return `k` as maxExtra, when LoE would allow `k+1`.

Reply:
Right, the current `computeLoEMaxExtra` is too conservative when the `loeFrag` is an extension of the chain we are switching to. We should change that :+1:

Reply:
`computeLoEMaxExtra` has been revised in 716d975.